libs/database/src/publish.rs

```diff
@@ -0,0 +1,273 @@
+  )
+  .fetch_one(executor)
+  .await?;
+
+  Ok(res.unwrap_or(false))
+}
+
+#[inline]
+pub async fn update_workspace_publish_namespace<'a, E: Executor<'a, Database = Postgres>>(
+  executor: E,
+  workspace_id: &Uuid,
+  new_namespace: &str,
+) -> Result<(), AppError> {
+  let res = sqlx::query!(
+    r#"
+      UPDATE af_workspace
+      SET publish_namespace = $1
+      WHERE workspace_id = $2
+    "#,
+    new_namespace,
+    workspace_id,
+  )
+  .execute(executor)
+  .await?;
+
+  if res.rows_affected() != 1 {
+    tracing::error!(
+      "Failed to update workspace publish namespace, workspace_id: {}, new_namespace: {}, rows_affected: {}",
+      workspace_id, new_namespace, res.rows_affected()
+    );
+  }
+
+  Ok(())
+}
+
+#[inline]
+pub async fn select_workspace_publish_namespace<'a, E: Executor<'a, Database = Postgres>>(
+  executor: E,
+  workspace_id: &Uuid,
+) -> Result<String, AppError> {
+  let res = sqlx::query_scalar!(
+    r#"
+      SELECT publish_namespace
+      FROM af_workspace
+      WHERE workspace_id = $1
+    "#,
+    workspace_id,
+  )
+  .fetch_one(executor)
+  .await?;
+
+  Ok(res)
+}
+
+#[inline]
+pub async fn insert_or_replace_publish_collabs<'a, E: Executor<'a, Database = Postgres>>(
+  executor: E,
+  workspace_id: &Uuid,
+  publisher_uuid: &Uuid,
+  publish_items: Vec<PublishCollabItem<serde_json::Value, Vec<u8>>>,
+) -> Result<(), AppError> {
+  let item_count = publish_items.len();
+  let mut view_ids: Vec<Uuid> = Vec::with_capacity(item_count);
+  let mut publish_names: Vec<String> = Vec::with_capacity(item_count);
+  let mut metadatas: Vec<serde_json::Value> = Vec::with_capacity(item_count);
+  let mut blobs: Vec<Vec<u8>> = Vec::with_capacity(item_count);
+  publish_items.into_iter().for_each(|item| {
+    view_ids.push(item.meta.view_id);
+    publish_names.push(item.meta.publish_name);
+    metadatas.push(item.meta.metadata);
+    blobs.push(item.data);
+  });
+
+  let res = sqlx::query!(
+    r#"
+      INSERT INTO af_published_collab (workspace_id, view_id, publish_name, published_by, metadata, blob)
+      SELECT * FROM UNNEST(
+        (SELECT array_agg((SELECT $1::uuid)) FROM generate_series(1, $7))::uuid[],
+        $2::uuid[],
+        $3::text[],
+        (SELECT array_agg((SELECT uid FROM af_user WHERE uuid = $4)) FROM generate_series(1, $7))::bigint[],
+        $5::jsonb[],
+        $6::bytea[]
+      )
+      ON CONFLICT (workspace_id, view_id) DO UPDATE
+      SET metadata = EXCLUDED.metadata,
+          blob = EXCLUDED.blob,
+          published_by = EXCLUDED.published_by,
+          publish_name = EXCLUDED.publish_name
+    "#,
+    workspace_id,
+    &view_ids,
+    &publish_names,
+    publisher_uuid,
+    &metadatas,
+    &blobs,
+    item_count as i32,
+  )
+  .execute(executor)
+  .await?;
+
+  if res.rows_affected() != item_count as u64 {
+    tracing::warn!(
+      "Failed to insert or replace publish collab meta batch, workspace_id: {}, publisher_uuid: {}, rows_affected: {}",
```

Comment:
This is a partial failure - some rows might have been updated, some might not:

1. If you want to be strongly consistent (either all collabs were duplicated or none of them are), you should rollback transaction at this point.
2. If you want to be elastic, we should probably return row_affected count and inform user how many collabs have we duplicated.

Also log both `{rows_affected}` and `{item_count}` so that we have a clue how many were untouched.

Reply:
@Horusiath iirc `{rows_affected}` account for both modified and inserted. my intent for every collab, it is either inserted or modified. I shall modify the code to log both  the `{rows_affected} and `{item_count}`, if they differ.